File: LeetCode/155MinStack/MinStack.py
class MinStack:

    def __init__(self):
        self.stack = []
        self.min = []

    # All operations -> Time:O(1)
    def push(self, val: int) -> None:
        self.stack.append(val)
        if not self.min or val < self.min[-1]:
            self.min.append(val)
        else:
            self.min.append(self.min[-1])


    def pop(self) -> None:
        self.stack.pop()
        self.min.pop()
        

    def top(self) -> int:
        return self.stack[-1]

        

    def getMin(self) -> int:
        return self.min[-1]

File: LeetCode/155MinStack/MinStack.cpp
class MinStack {
public:
    MinStack() {
    }
    
    vector<int> runStack;
    vector<int> minStack;
    
    void push(int val) {
        runStack.push_back(val);
        if (minStack.size() == 0 || val < minStack[minStack.size() - 1]){
            minStack.push_back(val);
        }
        else {
            minStack.push_back(minStack[minStack.size() - 1]);
        }
    }
    
    void pop() {
        minStack.erase(minStack.begin() + minStack.size() - 1);
        runStack.erase(runStack.begin() + runStack.size() - 1);
    }
    
    int top() {
        return runStack[runStack.size() - 1];
    }
    
    int getMin() {
        return minStack[minStack.size() - 1];
    }
};
